coreneuron/io/phase2.cpp

```diff
@@ -0,0 +1,1211 @@
+    actual_v = F.read_vector<double>(n_node);
+    if (n_diam > 0)
+        actual_diam = F.read_vector<double>(n_node);
+
+    auto& param_sizes = corenrn.get_prop_param_size();
+    auto& dparam_sizes = corenrn.get_prop_dparam_size();
+    for (size_t i = 0; i < n_mech; ++i) {
+        int type = types[i];
+        std::vector<int> nodeindices;
+        if (!corenrn.get_is_artificial()[type]) {
+            nodeindices = F.read_vector<int>(nodecounts[i]);
+        }
+        auto data = F.read_vector<double>(param_sizes[type] * nodecounts[i]);
+        std::vector<int> pdata;
+        if (dparam_sizes[type] > 0) {
+            pdata = F.read_vector<int>(dparam_sizes[type] * nodecounts[i]);
+        }
+        tmls.emplace_back(TML{nodeindices, data, pdata, 0, {}, {}});
+    }
+    output_vindex = F.read_vector<int>(nt.n_presyn);
+    output_threshold = F.read_vector<double>(n_real_output);
+    pnttype = F.read_vector<int>(nt.n_netcon - nrn_setup_extracon);
+    pntindex = F.read_vector<int>(nt.n_netcon - nrn_setup_extracon);
+    weights = F.read_vector<double>(n_weight);
+    delay = F.read_vector<double>(nt.n_netcon);
+    npnt = F.read_int();
+
+    for (size_t i = 0; i < n_mech; ++i) {
+        if (!corenrn.get_bbcore_read()[types[i]]) {
+            continue;
+        }
+        tmls[i].type = F.read_int();
+        int icnt = F.read_int();
+        int dcnt = F.read_int();
+        if (icnt > 0) {
+            tmls[i].iArray = F.read_vector<int>(icnt);
+        }
+        if (dcnt > 0) {
+            tmls[i].dArray = F.read_vector<double>(dcnt);
+        }
+    }
+
+    int n_vecPlayContinous = F.read_int();
+    vecPlayContinuous.reserve(n_vecPlayContinous);
+    for (size_t i = 0; i < n_vecPlayContinous; ++i) {
+        VecPlayContinuous2 item;
+        item.vtype = F.read_int();
+        item.mtype = F.read_int();
+        item.ix = F.read_int();
+        int sz = F.read_int();
+        item.yvec = F.read_vector<double>(sz);
+        item.tvec = F.read_vector<double>(sz);
+        vecPlayContinuous.push_back(item);
+    }
+
+    // store current checkpoint state to continue reading mapping
+    // The checkpoint numbering in phase 3 is a continuing of phase 2, and so will be restored
+    F.record_checkpoint();
+
+    if (F.eof())
+        return;
+
+    assert(F.read_int() == n_vecPlayContinous);
+
+    for (int i = 0; i < n_vecPlayContinous; ++i) {
+        auto &vecPlay = vecPlayContinuous[i];
+        vecPlay.last_index = F.read_int();
+        vecPlay.discon_index = F.read_int();
+        vecPlay.ubound_index = F.read_int();
+    }
+
+    pastim_index = F.read_int();
+
+    assert(F.read_int() == -1);
+
+    for (int i = 0; i < nt.n_presyn; ++i) {
+        preSynConditionEventFlags.push_back(F.read_int());
+    }
+
+    assert(F.read_int() == -1);
+    save_events(F);
+
+    assert(F.read_int() == -1);
+    save_events(F);
+}
+
+void Phase2::read_direct(int thread_id, const NrnThread& nt) {
```

Comment:
just to have things in the same order, can you make sure that in the phase1.cpp file the `read_direct` function also comes after the `read_file` - I know i'm nitpicking.. just trying to make things easier, since you're touching all this code.